Rockchip:nanopi-r2s improve boot failed

wevsty opened the pull request:
NanoPi-R2S boots with mmc init failed issue on some MicroSD cards.
Just like that.
```
U-Boot TPL 2020.07 (Aug 03 2020 - 21:34:34)
DDR4, 333MHz
BW=32 Col=10 Bk=4 BG=2 CS0 Row=15 CS=1 Die BW=16 Size=1024MB
Trying to boot from BOOTROM
Returning to boot ROM...

U-Boot SPL 2020.07 (Aug 03 2020 - 21:34:34 +0000)
Trying to boot from MMC1
spl: mmc init failed with error: -95
SPL: failed to boot from all boot devices
### ERROR ### Please RESET the board ###
```

After testing it was found that the initialization failed because the initialization wait time was too short, so I added startup-delay-us to fix the problem.
After my own NanoPi R2S test additions it has been able to boot properly from a specific MicroSD card.

Signed-off-by: Yuan Tao <ty@wevs.org>

wevsty committed:
uboot-rockchip: improve NanoPi-R2S some boot failed

Add regulator-boot-on parameters to vcc_sd.
Add startup-delay-us parameters to vcc_sdio.
This will improve the problem that in some SD cards are failing to boot up.

Signed-off-by: Yuan Tao <ty@wevs.org>

adrianschmutzler commented:
startup delay does not sound like a fix, but like a workaround for me?

Despite, you have some formal issues:
- Your commit title should contain a prefix ("rockchip:") and should be shorter, refer to https://openwrt.org/submitting-patches
- You should add a commit message to the patch, just like you explained the subject in your initial post.
- You should refer to the device in your commit title, as this only touches that particular device

blocktrron commented:
@wevsty Please specify the type of SD card you are using. I've tried multiple (20) SD cards from the past 15 years and none showed any issues.

wevsty commented:
> 
> 
> @wevsty Please specify the type of SD card you are using. I've tried multiple (20) SD cards from the past 15 years and none showed any issues.

The SD card I'm using is a Netac Extreme P500 Pro 32GB, and it has A1 U1 V10 printed on the card.
The card is marked on the vendor for use with video equipment such as a car recorder.
More information can be found on the manufacturer's website: http://www.netac.com/tf/316.html

I'm not the only one having problems with openwrt booting on some SD cards.
The results I tracked showed that mmc init failed error code -95 was due to the SD card not being able to respond in time when the controller sent commands to the SD card.

@adrianschmutzler
The error code -95 is represented in the source code as -EOPNOTSUPP.
An example is u-boot-2020.07/drivers/mmc/mmc.c
The error message is from mmc_init().
The final call chain for the error could be, mmc_init() -> mmc_start_init() -> mmc_get_op_cond() -> sd_send_op_cond()
```c
static int sd_send_op_cond(struct mmc *mmc, bool uhs_en)
{
	int timeout = 1000;
	int err;
	struct mmc_cmd cmd;

	while (1) {
		cmd.cmdidx = MMC_CMD_APP_CMD;
		cmd.resp_type = MMC_RSP_R1;
		cmd.cmdarg = 0;

		err = mmc_send_cmd(mmc, &cmd, NULL);

		if (err)
			return err;

		cmd.cmdidx = SD_CMD_APP_SEND_OP_COND;
		cmd.resp_type = MMC_RSP_R3;

		/*
		 * Most cards do not answer if some reserved bits
		 * in the ocr are set. However, Some controller
		 * can set bit 7 (reserved for low voltages), but
		 * how to manage low voltages SD card is not yet
		 * specified.
		 */
		cmd.cmdarg = mmc_host_is_spi(mmc) ? 0 :
			(mmc->cfg->voltages & 0xff8000);

		if (mmc->version == SD_VERSION_2)
			cmd.cmdarg |= OCR_HCS;

		if (uhs_en)
			cmd.cmdarg |= OCR_S18R;

		err = mmc_send_cmd(mmc, &cmd, NULL);

		if (err)
			return err;

		if (cmd.response[0] & OCR_BUSY)
			break;

		if (timeout-- <= 0)
			return -EOPNOTSUPP;

		udelay(1000);
	}

	if (mmc->version != SD_VERSION_2)
		mmc->version = SD_VERSION_1_0;

	if (mmc_host_is_spi(mmc)) { /* read OCR for spi */
		cmd.cmdidx = MMC_CMD_SPI_READ_OCR;
		cmd.resp_type = MMC_RSP_R3;
		cmd.cmdarg = 0;

		err = mmc_send_cmd(mmc, &cmd, NULL);

		if (err)
			return err;
	}

	mmc->ocr = cmd.response[0];

#if CONFIG_IS_ENABLED(MMC_UHS_SUPPORT)
	if (uhs_en && !(mmc_host_is_spi(mmc)) && (cmd.response[0] & 0x41000000)
	    == 0x41000000) {
		err = mmc_switch_voltage(mmc, MMC_SIGNAL_VOLTAGE_180);
		if (err)
			return err;
	}
#endif

	mmc->high_capacity = ((mmc->ocr & OCR_HCS) == OCR_HCS);
	mmc->rca = 0;

	return 0;
}
```
We can see that the reason for returning -EOPNOTSUPP is because we are waiting for the response timeout from the SD card.
Also as long as I add a DEBUG macro to U-Boot at compile time, the NanoPi R2S will start successfully.

Here's how it looks after adding the DEBUG macro
```
Trying to boot from MMC1
blk_find_device: if_type=6, devnum=0: mmc@ff500000.blk, 6, 0
ofnode_read_u32_index: vmmc-supply: x (75)
ofnode_read_bool: enable-active-high: false
ofnode_read_bool: regulator-boot-on: false
fdtdec_get_int: #gpio-cells: x (2)
size=x, ptr=18, limit=c0: 5000a8
size=x, ptr=18, limit=d8: 5000c0
0 -1
0 0
   - -1 -1 'gpio0@ff210000'
   - not found
0 0
   - -1 0 'pinctrl'
   - found
clk_set_defaults(gpio0@ff210000)
clk_set_default_parents: could not read assigned-clock-parents for 3fe8a0
ofnode_read_prop: assigned-clock-rates: <not found>
fdtdec_get_addr_size_auto_parent: na=2, ns=2, fdtdec_get_addr_size_fixed: reg: addr=00000000x
size=x, ptr=100, limit=1d8: 5000d8
size=x, ptr=6, limit=1de: 5001d8
ofnode_read_u32_index: startup-delay-us: (not found)
ofnode_read_u32_index: off-on-delay-us: (not found)
ofnode_read_u32_index: u-boot,off-on-delay-us: (not found)
0 -1
0 0
   - -1 -1 'sdmmc-regulator'
   - not found
0 -1
0 0
   - -1 -1 'pcfg-pull-up-4ma'
   - -1 -1 'sdmmc0-1'
   - -1 -1 'sdmmc0m1-gpio'
   - not found
0 0
   - -1 0 'pinctrl'
   - found
clk_set_defaults(sdmmc0-1)
clk_set_default_parents: could not read assigned-clock-parents for 3fea10
ofnode_read_prop: assigned-clock-rates: <not found>
0 -1
0 0
   - -1 -1 'pcfg-pull-up-4ma'
   - -1 0 'sdmmc0-1'
   - found
0 1
   - -1 -1 'pcfg-pull-up-4ma'
   - -1 0 'sdmmc0-1'
   - -1 -1 'sdmmc0m1-gpio'
   - not found
0 0
   - -1 0 'pinctrl'
   - found
clk_set_defaults(sdmmc0m1-gpio)
clk_set_default_parents: could not read assigned-clock-parents for 3feab0
ofnode_read_prop: assigned-clock-rates: <not found>
setting mux of GPIO0-30 to 0
setting pull of GPIO0-30 to 5
setting drive of GPIO0-30 to 4
ofnode_read_u32_index: regulator-min-microvolt: x (3300000)
ofnode_read_u32_index: regulator-max-microvolt: x (3300000)
ofnode_read_u32_index: regulator-init-microvolt: (not found)
ofnode_read_u32_index: regulator-min-microamp: (not found)
ofnode_read_u32_index: regulator-max-microamp: (not found)
ofnode_read_bool: regulator-always-on: false
ofnode_read_bool: regulator-boot-on: false
ofnode_read_u32_index: regulator-ramp-delay: (not found)
ofnode_find_subnode: regulator-state-mem: <none>
clk_set_defaults(sdmmc-regulator)
clk_set_default_parents: could not read assigned-clock-parents for 3feb80
ofnode_read_prop: assigned-clock-rates: <not found>
ofnode_read_u32_index: vqmmc-supply: x (30)
clock is disabled (0Hz)
Buswidth = 0, clock: 0
regulator_common_set_enable: dev='sdmmc-regulator', enable=0, delay=0, has_gpio=1
regulator_common_set_enable: done
regulator_common_set_enable: dev='sdmmc-regulator', enable=1, delay=0, has_gpio=1
regulator_common_set_enable: done
Buswidth = 1, clock: 0
clock is enabled (400000Hz)
Buswidth = 1, clock: 400000
Sending CMD0
Sending CMD8
Sending CMD55
Sending CMD41
Sending CMD55
Sending CMD41
Sending CMD55
Sending CMD41
Sending CMD55
Sending CMD41
Sending CMD55
Sending CMD41
Sending CMD55
Sending CMD41
Sending CMD55
Sending CMD41
Sending CMD55
Sending CMD41
Sending CMD2
Sending CMD3
Sending CMD9
Sending CMD7
Sending CMD55
Sending CMD51
Sending CMD55
Sending CMD6
Buswidth = 4, clock: 400000
clock is enabled (25000000Hz)
Buswidth = 4, clock: 25000000
clk_set_rate(clk=500000, rate=25000000)
spl: mmc boot mode: raw
blk_find_device: if_type=6, devnum=0: mmc@ff500000.blk, 6, 0
Sending CMD16
Sending CMD17
hdr read sector 4000, count=1
Found FIT
blk_find_device: if_type=6, devnum=0: mmc@ff500000.blk, 6, 0
Sending CMD16
Sending CMD18
Sending CMD12
fit read sector 4000, sectors=3, dst=1ff940, count=3, size=0x4a4
board_fit_config_name_match: rk3328-nanopi-r2s.dtb
Selecting config 'rk3328-nanopi-r2s.dtb'firmware: 'atf_1'
blk_find_device: if_type=6, devnum=0: mmc@ff500000.blk, 6, 0
Sending CMD16
Sending CMD18
Sending CMD12
External data: dst=40000, offset=b6f4c, size=17120
Image OS is ARM Trusted Firmware
```
ofnode_read_u32_index: startup-delay-us: (not found)
This message reminds me that booting up properly with DEBUG macros enabled could be because the SD card wasn't ready when the power was turned on, so the initialization reply to the SD card would eventually time out.

So here I added startup-delay-us. And now everything works.

I have more discussion on this issue again on the openwrt forum.
LINK : https://forum.openwrt.org/t/nanopi-r2s-is-a-great-openwrt-device/65374/169?u=wevsty

wevsty commented:
> 
> 
> startup delay does not sound like a fix, but like a workaround for me?
> 
> Despite, you have some formal issues:
> 
>     * Your commit title should contain a prefix ("rockchip:") and should be shorter, refer to https://openwrt.org/submitting-patches
> 
>     * You should add a commit message to the patch, just like you explained the subject in your initial post.
> 
>     * You should refer to the device in your commit title, as this only touches that particular device

I've made some changes to the commit, which should probably be the case?
Please let me know if you have any questions.

blocktrron commented:
I have a hard time understanding why this regulator should be the culprit, as it has to be turned on already for the SD card to operate.

I assume the card is not able to switch to low voltage mode within the specified timeframe after power-on, hence the delay on the SD power regulator works. In this case however, the delay should be more appropriately be applied to the regulator specifying the card voltage.

That being said, with a Samsung Evo 128G card, low-voltage operation works without an issue.

wevsty commented:
> 
> 
> I have a hard time understanding why this regulator should be the culprit, as it has to be turned on already for the SD card to operate.
> 
> I assume the card is not able to switch to low voltage mode within the specified timeframe after power-on, hence the delay on the SD power regulator works. In this case however, the delay should be more appropriately be applied to the regulator specifying the card voltage.
> 
> That being said, with a Samsung Evo 128G card, low-voltage operation works without an issue.
> 
> The website you've linked does not load / only shows some cryptic error in chinese.

Netac is a Chinese company, and I use a US server to access it, which is slow but works fine.

But that's not important, I also tried adding delay to vcc_sdio and the system still doesn't boot up properly.

I'm not a hardware developer, so I'm not sure exactly why.
All I can say is that this is the result of my testing.

wevsty commented:
> y be applied to the regulator specifying the card voltage.

I learned a little about SD Card.
By SD card bus
```
(1) Default Speed mode: 3.3V signaling, Frequency up to 25 MHz, up to 12.5 MB/sec
(2) High Speed mode: 3.3V signaling, Frequency up to 50 MHz, up to 25 MB/sec
(3) SDR12: UHS-I 1.8V signaling, Frequency up to 25 MHz, up to 12.5MB/sec
(4) SDR25: UHS-I 1.8V signaling, Frequency up to 50 MHz, up to 25MB/sec
(5) SDR50: UHS-I 1.8V signaling, Frequency up to 100 MHz, up to 50MB/sec
(6) SDR104: UHS-I 1.8V signaling, Frequency up to 208 MHz, up to 104MB/sec
(7) DDR50: UHS-I 1.8V signaling, Frequency up to 50 MHz, sampled on both clock edges, up to 50MB/sec
(8) UHS156: UHS-II RCLK Frequency Range 26MHz - 52MHz, up to 1.56Gbps per lane.
(9) UHS312: UHS-II RCLK Frequency Range 26MHz - 52MHz, up to 1.56Gbps two lanes half duplex 
```
Voltage adjustment will only be made if the SD card is controlled using the UHS mode.
But I don't see UHS support in the current device tree.
So presumably the SD card is always running in High Speed mode, and according to the startup log, my SD card reads at about 21.4MiB/s which is also consistent with this result.

bobafetthotmail commented:
there were similar issues in booting from some USB devices in kirkwood uboot (in Bodhi's Debian ARM for kirkwood, where USB boot is actually used) because the device took a slightly longer time to show up.

wevsty commented:
> there were similar issues in booting from some USB devices in kirkwood uboot (in Bodhi's Debian ARM for kirkwood, where USB boot is actually used) because the device took a slightly longer time to show up, and they had to increase timeouts.

I'm guessing the problem is because the sd card slot and the JTAG share pins.
According to the schematic diagram
SDMMC0_D2 is the same pin as JTAG_TCK_u.
SDMMC0_D3 is the same pin as JTAG_TMS_u.
Perhaps JTAG was trying to communicate when the vcc_sd power supply started, which gave the SD card controller an error.
The vcc_sd delay to start powering up prevents the SD card from reading the wrong data.

I've seen others raise similar questions.

> Leaving the switching function enabled did result in mmc timeouts and rcu stalls thus hanging the system on 3.19-rc1.

LINK: https://patchwork.kernel.org/patch/5607321/

That would make a lot of sense.


I've also added a delay to vcc_sdio. Since @jayanta525 was experiencing some compatibility issues with some SD cards, he thought this would help improve compatibility.
I didn't think it would cause any problems either, so I added it together.

qhlai commented:
i have the same problem ，i have two card
 sandisk ultra 32gb A1 C10
 `
U-Boot TPL 2020.07 (Aug 04 2020 - 13:51:08)
DDR4, 333MHz
BW=32 Col=10 Bk=4 BG=2 CS0 Row=15 CS=1 Die BW=16 Size=1024MB
Trying to boot from BOOTROM
Returning to boot ROM...

U-Boot SPL 2020.07 (Aug 04 2020 - 13:51:08 +0000)
Trying to boot from MMC1
mmc_load_image_raw_sector: mmc block read error
SPL: failed to boot from all boot devices
### ERROR ### Please RESET the board ###
`
 kingston 16GB C16G （is a old card and i am not sure is a real kingston card） show
 `
U-Boot SPL 2020.07 (Aug 04 2020 - 13:51:08 +0000)
Trying to boot from MMC1
spl: mmc init failed with error: -95
SPL: failed to boot from all boot devices
### ERROR ### Please RESET the board ###
`
maybe the boot is not really well ，because when i use Friendlywrt, they all can boot successfully.

wevsty commented:
> 
> 
> i have the same problem ，i have two card
> sandisk ultra 32gb A1 C10
> `U-Boot TPL 2020.07 (Aug 04 2020 - 13:51:08)`
> `DDR4, 333MHz`
> `BW=32 Col=10 Bk=4 BG=2 CS0 Row=15 CS=1 Die BW=16 Size=1024MB`
> `Trying to boot from BOOTROM`
> `Returning to boot ROM...`
> ``
> `U-Boot SPL 2020.07 (Aug 04 2020 - 13:51:08 +0000)`
> `Trying to boot from MMC1`
> `mmc_load_image_raw_sector: mmc block read error`
> `SPL: failed to boot from all boot devices`
> `### ERROR ### Please RESET the board ###`
> kingston 16GB C16G （is a old card and i am not sure is a real kingston card） show
> 
> `U-Boot SPL 2020.07 (Aug 04 2020 - 13:51:08 +0000)`
> `Trying to boot from MMC1`
> `spl: mmc init failed with error: -95`
> `SPL: failed to boot from all boot devices`
> `### ERROR ### Please RESET the board ###`
> 
> maybe the boot is not really well ，because when i use Friendlywrt, they all can boot successfully.

Has it improved since trying the patch here?

qhlai commented:
> 
> 
> > i have the same problem ，i have two card
> > sandisk ultra 32gb A1 C10
> > `U-Boot TPL 2020.07 (Aug 04 2020 - 13:51:08)`
> > `DDR4, 333MHz`
> > `BW=32 Col=10 Bk=4 BG=2 CS0 Row=15 CS=1 Die BW=16 Size=1024MB`
> > `Trying to boot from BOOTROM`
> > `Returning to boot ROM...`
> > ``
> > `U-Boot SPL 2020.07 (Aug 04 2020 - 13:51:08 +0000)`
> > `Trying to boot from MMC1`
> > `mmc_load_image_raw_sector: mmc block read error`
> > `SPL: failed to boot from all boot devices`
> > `### ERROR ### Please RESET the board ###`
> > kingston 16GB C16G （is a old card and i am not sure is a real kingston card） show
> > `U-Boot SPL 2020.07 (Aug 04 2020 - 13:51:08 +0000)`
> > `Trying to boot from MMC1`
> > `spl: mmc init failed with error: -95`
> > `SPL: failed to boot from all boot devices`
> > `### ERROR ### Please RESET the board ###`
> > maybe the boot is not really well ，because when i use Friendlywrt, they all can boot successfully.
> 
> Has it improved since trying the patch here?

我还没试，太晚了

wevsty commented:
> > 懒得写英文了，中文打算了
> > 我还没试，太晚了，我对这个项目还不是很熟悉，怎么改我还得看，头疼，朗科京东联名卡也不行？
> > 然后我试过一个骚操作，卡住的时候热插拔一次就能成功进入ext4，对我那张金士顿的卡，不过这是另一个问题，是总线速率协调的问题，我的金士顿卡刷squahfs和你的问题完全一样。
> > 但是闪迪那张ext4和squashfs都不行
> > 之前也查过可能是cmd不兼容，但我觉得增加延时的方案可行
> > 总之这个boot问题好多，能不能把friendlywrt的boot直接移过来？但我又找不到它的boot文件。太难为我这个大二非计算机学生了

建议使用英文，因为还有其他的开发者，方便他们直接查阅。
OpenWrt也是有参考friendlywrt的来适配的，直接搬friendlywrt的u-boot应该不太可行。
由于我没有其他的SD卡可以测试了，所以需要一些反馈。


Translated:
English is recommended, as there are other developers can direct viewing.
OpenWrt also has a reference to friendlywrt, and it shouldn't be feasible to move friendlywrt's u-boot directly.
Since I don't have any more SD cards to test with, I need some feedback.

qhlai commented:
> 建议使用英文，因为还有其他的开发者，方便他们直接查阅。
> OpenWrt也是有参考friendlywrt的来适配的，直接搬friendlywrt的u-boot应该不太可行。
> 由于我没有其他的SD卡可以测试了，所以需要一些反馈。
> 如果你不想自己编译，可以给我发邮件。
> 我可以提供我自己编译的版本。
> 
> Translated:
> English is recommended, as there are other developers can direct viewing.
> OpenWrt also has a reference to friendlywrt, and it shouldn't be feasible to move friendlywrt's u-boot directly.
> Since I don't have any more SD cards to test with, I need some feedback.
> If you don't want to compile it yourself, you can send me an email.
> I can provide the version I compiled myself.

OK, using your system is the most convient way to test it, and i want to know your email address,or you can send it to my email 632489313@qq.com

wevsty commented:
> OK, using your system is the most convient way to test it, and i dont know your email address,or you can send it to my email at 

Please check your e-mail.

qhlai commented:
> 
> 
> > OK, using your system is the most convient way to test it, and i dont know your email address,or you can send it to my email at
> 
> Please check your e-mail.

sandisk ultra 32GB A1 C10 has the same problem,  error 53
```
U-Boot TPL 2020.07 (Aug 04 2020 - 13:51:08)
DDR4, 333MHz
BW=32 Col=10 Bk=4 BG=2 CS0 Row=15 CS=1 Die BW=16 Size=1024MB
Trying to boot from BOOTROM
Returning to boot ROM...

U-Boot SPL 2020.07 (Aug 04 2020 - 13:51:08 +0000)
Trying to boot from MMC1
mmc_load_image_raw_sector: mmc block read error
SPL: failed to boot from all boot devices
### ERROR ### Please RESET the board ###
```

but kingston tf card  may become bette,dont disappera `spl: mmc init failed with error: -95`,but still has problem which also appear in friendlywrt,it might be the bus speed problem

boot stop here
```
    0.603091] rk808-rtc rk808-rtc: setting system clock to 2016-01-21T08:50:09 UTC (1453366209)
[    0.606096] Waiting for root device PARTUUID=5452574f-02...
[    0.726061] usb 4-1: new SuperSpeed Gen 1 USB device number 2 using xhci-hcd
[    1.812472] mmc0: card never left busy state
[    1.812872] mmc0: error -110 whilst initialising SD card
[    1.828230] mmc_host mmc0: Bus speed (slot 0) = 300000Hz (slot req 300000Hz, actual 300000HZ div = 0)
[    3.195411] mmc0: card never left busy state
[    3.195802] mmc0: error -110 whilst initialising SD card
[    3.211338] mmc_host mmc0: Bus speed (slot 0) = 200000Hz (slot req 200000Hz, actual 200000HZ div = 0)
[    4.622590] mmc0: card never left busy state
[    4.622981] mmc0: error -110 whilst initialising SD card
[    4.639075] mmc_host mmc0: Bus speed (slot 0) = 100000Hz (slot req 100000Hz, actual 100000HZ div = 0)
[    6.181176] mmc0: card never left busy state
[    6.181566] mmc0: error -110 whilst initialising SD card
[   10.166855] mmc_host mmc0: Bus speed (slot 0) = 400000Hz (slot req 400000Hz, actual 400000HZ div = 0)
[   10.362441] mmc_host mmc0: Bus speed (slot 0) = 50000000Hz (slot req 50000000Hz, actual 50000000HZ div = 0)
```
but if i hot plug out and in, it can run and enter the system well 
```
[   10.363388] mmc0: new high speed SDHC card at address 1234
[   10.365537] mmcblk0: mmc0:1234 SA16G 14.6 GiB 
[   10.368401]  mmcblk0: p1 p2
[   10.389308] VFS: Mounted root (squashfs filesystem) readonly on device 179:2.
[   10.391169] Freeing unused kernel memory: 1664K
[   10.410024] Run /sbin/init as init process
[   10.649823] init: Console is alive
[   10.929670] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[   10.975638] ehci-fsl: Freescale EHCI Host controller driver
[   10.978493] uhci_hcd: USB Universal Host Controller Interface driver
[   10.986160] usbcore: registered new interface driver uas
[   10.986926] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
[   10.994265] init: - preinit -
[   11.400280] random: jshn: uninitialized urandom read (4 bytes read)
[   11.446938] random: jshn: uninitialized urandom read (4 bytes read)
[   11.466912] random: hexdump: uninitialized urandom read (5 bytes read)
Press the [f] key and hit [enter] to enter failsafe mode
Press the [1], [2], [3] or [4] key and hit [enter] to select the debug level
```

wevsty commented:
> mmc_load_image_raw_sector: mmc block read error

This looks like a problem with the SD card file system.
I would suggest performing a full low level formatting and then try again, if it still errors then I think I need to do more investigation.

> [    1.812472] mmc0: card never left busy state

This should be a hint from the Linux kernel, which U-Boot has in fact successfully booted.

qhlai commented:
> 
> 
> > mmc_load_image_raw_sector: mmc block read error
> 
> This looks like a problem with the SD card file system.
> I would suggest performing a full low level formatting and then try again.
> You can use this tool to perform the formatting.
> Download link https://www.sdcard.org/downloads/formatter/
> If it still errors then I think I need to do more investigation.
> 
> > [    1.812472] mmc0: card never left busy state
> 
> This should be a hint from the Linux kernel, which U-Boot has in fact successfully booted.

yesterday ,i have use diskgenuins to write  00 to fill all the block,and win10's format.(not quick format ) .
it's no difference

wevsty commented:
> 
> 
> > > mmc_load_image_raw_sector: mmc block read error
> > 
> > 
> > This looks like a problem with the SD card file system.
> > I would suggest performing a full low level formatting and then try again.
> > You can use this tool to perform the formatting.
> > Download link https://www.sdcard.org/downloads/formatter/
> > If it still errors then I think I need to do more investigation.
> > > [    1.812472] mmc0: card never left busy state
> > 
> > 
> > This should be a hint from the Linux kernel, which U-Boot has in fact successfully booted.
> 
> yesterday ,i have use diskgenuins to write 00 to fill all the block,and win10's format.(not quick format ) .
> it's no difference

I suggest defining these macros for u-boot and recompile.
If you still get a startup error, post the log again.
#define DEBUG
#define CONFIG_MMC_TRACE
#define CONFIG_SUPPORT_EMMC_BOOT

If you need additional assistance, please contact

qhlai commented:
> 
> 
> > > > mmc_load_image_raw_sector: mmc block read error
> > > 
> > > 
> > > This looks like a problem with the SD card file system.
> > > I would suggest performing a full low level formatting and then try again.
> > > You can use this tool to perform the formatting.
> > > Download link https://www.sdcard.org/downloads/formatter/
> > > If it still errors then I think I need to do more investigation.
> > > > [    1.812472] mmc0: card never left busy state
> > > 
> > > 
> > > This should be a hint from the Linux kernel, which U-Boot has in fact successfully booted.
> > 
> > 
> > yesterday ,i have use diskgenuins to write 00 to fill all the block,and win10's format.(not quick format ) .
> > it's no difference
> 
> I suggest defining these macros for u-boot and recompile.
> If you still get a startup error, post the log again.
> #define DEBUG
> #define CONFIG_MMC_TRACE
> #define CONFIG_SUPPORT_EMMC_BOOT
> 
> If you need additional assistance, please contact

Okay ,very thx.
I try to compile one , I clone one project from `coolsnowwolf/lede`,and replace the file from which your provided in anothor discuss. I am not sure if I can use  `coolsnowwolf/lede` or I should clone offical project.
I wonder which tf card is compatible with this device， maybe buy a new one is the most effcient way to solve it。（laugh）

wevsty commented:
Following openwrt's patching style, I've changed the changes to a separate patch.

I've also added code to disable JTAG switching for RK3328.
The code references : https://github.com/rockchip-linux/u-boot/blob/next-dev/arch/arm/mach-rockchip/rk3328/rk3328.c
Disabling JTAG switching has no effect on my SD card, but I think it's good for SD card compatibility.

wevsty commented:
Maybe someone doesn't like the option of adding startup-delay-us.

I have modified the solution and my tests have shown that adding regulator-boot-on to vcc_sd also solves the problem.

I guess the problem is because it takes some time to adjust the voltage, and this time is different for different SD cards, and the SD card may not be ready to receive the command when the CPU initializes it.

The example provided at rockchip does the same thing.
LINK: https://github.com/rockchip-linux/u-boot/blob/next-dev/arch/arm/dts/rk3328-evb.dts

So I tweaked the submitted patch.
I think this solution should be acceptable.

@blocktrron Are there any plans to merge this patch? Or is there something I'm not doing right?

1715173329 commented:
@wevsty Maybe you need squash all your commits into one?

wevsty commented:
> 
> 
> @wevsty Maybe you need squash all your commits into one?

already squash all the commits.

jayanta525 commented on package/boot/uboot-rockchip/patches/101-rockchip-rk3328-nanopi-r2s-improve-boot-failed.patch:
```diff
@@ -0,0 +1,47 @@
+ CONFIG_FASTBOOT_BUF_ADDR=0x800800
+ CONFIG_FASTBOOT_CMD_OEM_FORMAT=y
+ CONFIG_ROCKCHIP_GPIO=y
++CONFIG_REGULATOR_GPIO=y
```
This is not required in uboot-defconfig.

wevsty replied:
OK, This has been deleted.

wevsty commented:
This patch is included in the master branch.
So close PR.
